server/chat-plugins/logviewer.ts

```diff
@@ -0,0 +1,137 @@
+import {FS} from '../../lib/fs';
+
+interface RoomlogReader {
+	readLogs: (roomid: RoomID, year: string, month: string, date: string) => string;
+	months: (room: Room) => string[][];
+}
+
+class RoomlogReaderFS implements RoomlogReader {
+	readLogs(roomid: RoomID, year: string, month: string, date: string) {
+		let buf;
+		try {
+			buf = FS(`logs/chat/${roomid}/${year}-${month}/${year}-${month}-${date}.txt`).readIfExistsSync();
+		} catch (e) {
+			buf = '';
+		}
+		return buf;
+	}
+	months(room: Room) {
+		const months: string[] = [];
+		const days: string[] = [];
+		for (const m of FS(`logs/chat/${room}/`).readdirSync()) {
+			if (!FS(`logs/chat/${room}/${m}/`).isDirectorySync()) continue;
+			months.push(m);
+		}
+		for (const month of months) {
+			for (const day of FS(`logs/chat/${room}/${month}`).readdirSync()) {
+				if (day.endsWith('.txt')) days.push(day.slice(0, -4));
+			}
+		}
+		return [months, days];
+	}
+}
+
+class RoomlogViewer<T extends RoomlogReader> {
+	reader: T;
+	constructor(reader: T) {
+		this.reader = reader;
+	}
+	button(datestring: string, room: Room, nodate = false) {
+		if (!nodate) {
+			return `<button class="button" name="send" value="/join view-logs--${room}--${datestring}">${datestring}</button>`;
+		} else {
+			return `<button class="button" name="send" value="/join view-logs--${room}">${datestring}</button>`;
+		}
+	}
+	structure(room: Room) {
+		let buf = `<b>Logs for ${room}</b><br>`;
+		const [months, days] = this.reader.months(room);
+		for (const month of months) {
+			buf += `<br><br><b>${month}</b><br>`;
+			for (const day of days) {
+				if (!day.includes(month)) continue;
+				buf += this.button(day, room);
+			}
+		}
+		return buf;
+	}
+	clean(content: string, showall = false) {
+		const cleaned = [];
+		for (let line of content.split('\n')) {
+			const part = line.split('|');
+			const cp = ['c', 'c:'];
+			if (cp.includes(part[1])) {
+				line = `[${part[0].trim()}] ${part[2]}: ${part[3]}`;
+				cleaned.push(`&nbsp;${line}`);
+			} else if (part[1] === 'html') {
+				line = `[${part[0].trim()}]: ${part[2]}`;
+				cleaned.push(`&nbsp;${line}`);
+			} else if (showall) {
+				cleaned.push(`&nbsp;${line}`);
+			} else {
+				const exceptions = ['N', 'J', 'L', 'userstats'];
+				if (!exceptions.includes(part[1])) cleaned.push(`&nbsp;${line}`);
+			}
+		}
+		return cleaned;
+	}
+}
+
+const viewer = new RoomlogViewer(new RoomlogReaderFS());
+
+export const commands: ChatCommands = {
+	viewlogs(target, room, user) {
+		if (!room.chatRoomData || room.roomid.includes('-')) return this.errorReply("This room has no logs.");
```

Comment:
Help tickets contain hyphens and they do have logs. TBH, I don't really see the use of this check, if the room doesn't have logs they can see that it doesn't in the HTML page.